src/Controllers/RegistrationController.php

```diff
@@ -0,0 +1,268 @@
+        SessionInterface $session,
+        Authenticator $auth,
+        EngelsystemMailer $mail,
+        LoggerInterface $log,
+        Config $config
+    ) {
+        $this->response = $response;
+        $this->session = $session;
+        $this->auth = $auth;
+        $this->mail = $mail;
+        $this->log = $log;
+        $this->config = $config;
+    }
+
+    /**
+     * @return Response
+     */
+    public function register(): Response
+    {
+        return $this->showForm();
+    }
+
+    /**
+     * @param Request $request
+     * @return Response
+     */
+    public function postRegister(Request $request): Response
+    {
+        $isPending = $this->config->get('enable_pending_registrations') && $this->auth->user() != null;
+
+        if (!($isPending || $this->config->get('registration_enabled'))) {
+            error(__('registration.disabled'));
+
+            return $this->response->redirectTo('/');
+        }
+
+        $rules = [
+            'login'     => 'required|between:1:23|regex:/^([\p{L}\p{N}\-_. ]+)$/ui',
+            'email'     => 'required|email',
+            'mobile'    => 'optional|phone',
+        ];
+
+        if ($this->config->get('enable_tshirt_size')) {
+            $rules = array_merge($rules, [
+                'tshirt_size' => 'required|in:' . join(',', array_keys($this->config->get('tshirt_sizes'))),
+            ]);
+        }
+        if ($this->config->get('enable_planned_arrival')) {
+            $buildup = $this->config->get('buildup_start');
+            $teardown = $this->config->get('teardown_end');
+            $boundsRule = '';
+            if (!empty($buildup) && !empty($teardown)) {
+                $boundsRule = '|between:' . $buildup . ':' . $teardown;
+            } elseif (!empty($buildup)) {
+                $boundsRule = '|min:' . $buildup;
+            } elseif (!empty($teardown)) {
+                $boundsRule = '|max:' . $teardown;
+            }
+            $rules = array_merge($rules, ['planned_arrival_date' => 'required|date:Y-m-d' . $boundsRule]);
+        }
+        if (!$isPending) {
+            $rules = array_merge($rules, [
+                'password'              => 'required|min:' . $this->config->get('min_password_length'),
+                'password_confirmation' => 'required',
+            ]);
+        }
+        if ($this->config->get('enable_dect')) {
+            $rules = array_merge($rules, ['dect'  => 'optional|length:0:40']);
+        }
+        if ($this->config->get('enable_user_name')) {
+            $rules = array_merge($rules, [
+                'firstname'   => 'optional|alpha',
+                'lastname'  => 'optional|alpha',
+            ]);
+        }
+
+        $data = $this->validate($request, $rules);
+
+        if (User::whereName($data['login'])->count() > 0) {
+            $this->session->set('errors', array_merge($this->session->get('errors', []), ['registration.exists.login']));
+
+            return $this->showForm();
```

Comment:
Returning that early means that the user has to send the form multiple times, a better way would be to run all checks and after that show the page again if any check fails